taf/testlib/virtual_env.py

```diff
@@ -325,25 +335,29 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
 
         return False
 
-    def _get_external_elements(self, routers_client=None):
+    def _get_external_elements(self, routers_client=None, tenant_id=None):
         if not routers_client:
             routers_client = self.handle.os_adm.routers_client
 
         net_filter = {'router:external': True}
         nets = [net['id'] for net in self.handle._list_networks(**net_filter)]
-        router_2_net_map = {}
+        net_2_router_map = {}
         routers_resp = routers_client.list_routers()
         for router in routers_resp['routers']:
-            ext_gw_info = router.get('external_gateway_info')
-            if ext_gw_info:
-                net_id = ext_gw_info.get('network_id')
-                if net_id and net_id in nets:
-                    router_2_net_map[router['id']] = net_id
-
-        return router_2_net_map
+            if tenant_id and tenant_id == router.get('tenant_id'):
+                ext_gw_info = router.get('external_gateway_info')
+                if ext_gw_info:
+                    net_id = ext_gw_info.get('network_id')
+                    if net_id and net_id in nets:
+                        net_2_router_map[net_id] = router['id']
```

Comment:
I missed the fact that `net` in my suggestion is only used for the conditional expression, so we don't need the `net = nets.get('net_id')` and the conditional expression can be as it has been: `if net_id and net_id in nets`